SC_redesign/Crop_and_Soil/crop/leaf_area_index.py

```diff
@@ -0,0 +1,204 @@
+from math import exp, log, sqrt
+
+"""
+This module is based off of the 'Canopy Cover and Height' section of SWAT
+"""
+
+class LeafAreaIndex:
+    def __init__(self):
+        # fixed attributes (unchanged during simulations)
+        self.first_heat_fraction_point = 0.15
+        self.second_heat_fraction_point = 0.50
+        self.first_leaf_fraction_point = 0.01
+        self.second_leaf_fraction_point = 0.95
+        self.max_canopy_height = 2.5  # m
+        self.growth_factor = 1.0
+        self.max_leaf_area_index = 3.0
+        self.senescent_heat_fraction = 0.9
+        # variable attributes (change throughout simulations)
+        self.leaf_area_index = 0
+        self.heat_fraction = 0.73
```

Comment:
@ew3361zh, one question here. I've gone back and forth about documenting attributes in each of the inherited parent classes or in the eventual child class `Crop`. On one hand, documenting the arguments in parents would be very convenient and would prevent the need to re-declare all attributes in the child. On the other hand, some arguments (e.g., `heat_fraction`) are used by multiple parent classes, so the risk of duplicating documentation and having conflicting documentation for the same attribute is higher.

```suggestion
        self.heat_fraction = 0.73
        """total heat units accumulated by the plant, as a proportion of potential heat units"""
```

What do you think?

Reply:
I believe there are ways to cross reference docstrings between parent and child classes with Sphinx documentation and there are also ways to write decorator functions that you can use to manually accomplish the same thing but I don't have a lot of experience with either unfortunately and I'm not sure how far down that path you want to go. Maybe a good question for Pooya?

This is the example of the decorator function I've seen pop up a few times.
https://softwareengineering.stackexchange.com/a/386758

Reply:
Hi Clay, can you clarify on what you meant by 'some arguments (e.g., heat_fraction) are used by multiple parent classes'?

Reply:
Definitely: The main class `Crop` will inherit from `LeafAreaIndex` and another class `HeatUnits` (in fact, more than just these 2). Both parent classes contain processes (methods) that use the `heat_fraction` attribute. Neither of the parent classes are likely to be used by anything - the child `Crop` will almost always be used instead. So the question becomes: where do we document `heat_fraction`? In all parents? In just one of the parent classes, with a cross-reference in the others? Or in the main `Crop` class?

Here's a short example:

```python
# ---- leaf_area_index.py ----
class LeafAreaIndex:
  def __init__(self):
    self.heat_fraction = 0.2
    """fraction of potential heat units accumulated by a plant to date"""
  def leaf_method(self):
    print(f"do something with {self.heat_fraction}")
  # ... more code ...

# ---- heat_units.py ----
class HeatUnits:
  def __init__(self):
    self.heat_fraction = 0.2
    """fraction of potential heat units accumulated by a plant to date"""
  def heat_method(self):
    print(f"do something different with {self.heat_fraction}")
  # ... more code ...

# ---- crop.py ----
class Crop(LeafAreaIndex, HeatUnits):
  def __init__(self, heat_fraction):
    LeafAreaIndex.__init__(self)
    HeatUnits.__init__(self)

    self.heat_fraction = heat_fraction
    """fraction of potential heat units accumulated by a plant to date"""
```

It seems to me like documenting the attribute 3+ times is bad and will lead to contradictions.